src/Init/Data/String/Basic.lean

```diff
@@ -332,44 +326,52 @@ namespace Substring
 @[inline] def toIterator : Substring → String.Iterator
   | ⟨s, b, _⟩ => ⟨s, b⟩
 
+/-- Returns the codepoint at the given offset into the substring. -/
 @[inline] def get : Substring → String.Pos → Char
   | ⟨s, b, _⟩, p => s.get (b+p)
 
+/-- Given an offset of a codepoint into the substring,
+returns the offset there of the next codepoint. -/
 @[inline] def next : Substring → String.Pos → String.Pos
   | ⟨s, b, e⟩, p =>
-    let p := s.next (b+p)
-    if p > e then e - b else p - b
+    let absP := b+p
+    if absP = e then p else s.next absP - b
 
-@[inline] def prev : Substring → String.Pos → String.Pos
+/-- Given an offset of a codepoint into the substring,
+returns the offset there of the previous codepoint. -/
+@[inline] private def prev : Substring → String.Pos → String.Pos
   | ⟨s, b, _⟩, p =>
-    if p = b then p else s.prev (b+p) - b
```

Comment:
I think this API was broken since `p = b` assumes `p` is an offset into the underlying string `s` but then we do `b+p`. With this change both `next` and `prev` take an offset into the substring. I also made some functions private since they could be confusing and aren't used anywhere.